Decoding a pointer is a fatal error - should we just warn?

If GitLFS cannot decode a pointer file then it fails. This can happen if a user tracks `*.bin` files on a case *sensitive* file system. GitLFS would not track `*.BIN` files on that system. If another users works with the repo on a case *insensitive* file system then GitLFS would try to decode pointers for the `*.BIN` which would fail.  I wonder if it should rather just warn the user in that case? 

Plus the failure is different for large and small files (@ttaylorr I assume this is because we read the first 1024 bytes of the pointer, right?).

## Testcase:

```
git version 2.11.0
git-lfs/1.5.2 (GitHub; darwin amd64; go 1.7.3)
```

Error message if decoding of a large file fails:
```
$ git clone --branch largefile git@github.com:larsxschneider/lfstest-individual-files.git smudge-error
Cloning into 'smudge-error'...
remote: Counting objects: 33, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 33 (delta 2), reused 0 (delta 0), pack-reused 21
Receiving objects: 100% (33/33), 6.21 KiB | 0 bytes/s, done.
Resolving deltas: 100% (5/5), done.
fatal: big-file1.bin: smudge filter lfs failed
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0x3a1400, 0xc420010250)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
github.com/git-lfs/git-lfs/git.(*FilterProcessScanner).readRequest(0xc420172fc0, 0xc4200f51f0, 0x0, 0xc420051a10)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/git/filter_process_scanner.go:181 +0x218
github.com/git-lfs/git-lfs/git.(*FilterProcessScanner).Scan(0xc420172fc0, 0x402e11)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/git/filter_process_scanner.go:143 +0x50
github.com/git-lfs/git-lfs/commands.filterCommand(0xc4200e9380, 0x602ba0, 0x0, 0x0)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/commands/command_filter_process.go:76 +0x1f9
github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra.(*Command).execute(0xc4200e9380, 0x602ba0, 0x0, 0x0, 0xc4200e9380, 0x602ba0)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra/command.go:477 +0x170
github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4200e84e0, 0xc420051ed0, 0x1)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra/command.go:551 +0x380
github.com/git-lfs/git-lfs/commands.Run()
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/commands/run.go:65 +0x196
main.main()
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/git-lfs.go:33 +0x111
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

Error message if decoding of a small file fails:
```
$ git clone --branch smallfile git@github.com:larsxschneider/lfstest-individual-files.git smudge-error
Cloning into 'smudge-error'...
remote: Counting objects: 33, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 33 (delta 2), reused 0 (delta 0), pack-reused 21
Receiving objects: 100% (33/33), 6.21 KiB | 0 bytes/s, done.
Resolving deltas: 100% (5/5), done.
fatal: big-file1.bin: smudge filter lfs failed
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

Plus, even if we disable the smudge filter, the checkout fails:
```
$ GIT_LFS_SKIP_SMUDGE=1 git clone --branch largefile git@github.com:larsxschneider/lfstest-individual-files.git smudge-error
Cloning into 'smudge-error'...
remote: Counting objects: 33, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 33 (delta 2), reused 0 (delta 0), pack-reused 21
Receiving objects: 100% (33/33), 6.21 KiB | 0 bytes/s, done.
Resolving deltas: 100% (5/5), done.
fatal: big-file1.bin: smudge filter lfs failed
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0x3a1400, 0xc420010250)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
github.com/git-lfs/git-lfs/git.(*FilterProcessScanner).readRequest(0xc420135dc0, 0xc420106a30, 0x0, 0xc420051a10)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/git/filter_process_scanner.go:181 +0x218
github.com/git-lfs/git-lfs/git.(*FilterProcessScanner).Scan(0xc420135dc0, 0x402e11)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/git/filter_process_scanner.go:143 +0x50
github.com/git-lfs/git-lfs/commands.filterCommand(0xc4200f8000, 0x602ba0, 0x0, 0x0)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/commands/command_filter_process.go:76 +0x1f9
github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra.(*Command).execute(0xc4200f8000, 0x602ba0, 0x0, 0x0, 0xc4200f8000, 0x602ba0)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra/command.go:477 +0x170
github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42008d040, 0xc420051ed0, 0x1)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra/command.go:551 +0x380
github.com/git-lfs/git-lfs/commands.Run()
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/commands/run.go:65 +0x196
main.main()
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/git-lfs.go:33 +0x111
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

--- a/commands/command_filter_process.go
+++ b/commands/command_filter_process.go
@@ -4,7 +4,9 @@ import (
 	"fmt"
 	"io"
 	"os"
+	"strings"
 
+	"github.com/git-lfs/git-lfs/errors"
 	"github.com/git-lfs/git-lfs/filepathfilter"
 	"github.com/git-lfs/git-lfs/git"
 	"github.com/git-lfs/git-lfs/lfs"
@@ -45,6 +47,8 @@ func filterCommand(cmd *cobra.Command, args []string) {
 	skip := filterSmudgeSkip || cfg.Os.Bool("GIT_LFS_SKIP_SMUDGE", false)
 	filter := filepathfilter.New(cfg.FetchIncludePaths(), cfg.FetchExcludePaths())
 
+	var malformed []string
+
 	for s.Scan() {
 		var err error
 		var w *git.PktlineWriter
@@ -64,6 +68,11 @@ func filterCommand(cmd *cobra.Command, args []string) {
 			ExitWithError(fmt.Errorf("Unknown command %q", req.Header["command"]))
 		}
 
+		if errors.IsNotAPointerError(err) {
+			malformed = append(malformed, req.Header["pathname"])
+			err = nil
+		}
+
 		var status string
 		if ferr := w.Flush(); ferr != nil {
 			status = statusFromErr(ferr)
@@ -72,6 +81,11 @@ func filterCommand(cmd *cobra.Command, args []string) {
 		}
 
 		s.WriteStatus(status)
+	}
+
+	if len(malformed) > 0 {
+		fmt.Fprintf(os.Stderr, "Encountered %d files that should have been pointers, but weren't:\n", len(malformed))
+		fmt.Fprintf(os.Stderr, "%s\n", strings.Join(malformed, ", "))
 	}
 
 	if err := s.Err(); err != nil && err != io.EOF {

--- a/commands/command_filter_process.go
+++ b/commands/command_filter_process.go
@@ -83,7 +83,7 @@ func filterCommand(cmd *cobra.Command, args []string) {
 	}
 
 	if len(malformed) > 0 {
-		fmt.Fprintf(os.Stderr, "Encountered %d files that should have been pointers, but weren't:\n", len(malformed))
+		fmt.Fprintf(os.Stderr, "Encountered %d file(s) that should have been pointers, but weren't:\n", len(malformed))
 		for _, m := range malformed {
 			fmt.Fprintf(os.Stderr, "\t%s\n", m)
 		}

--- a/commands/command_filter_process.go
+++ b/commands/command_filter_process.go
@@ -1,7 +1,6 @@
 package commands
 
 import (
-	"bytes"
 	"fmt"
 	"io"
 	"os"
@@ -29,35 +28,6 @@ const (
 // dictating whether or not to skip the smudging process, leaving pointers as-is
 // in the working tree.
 var filterSmudgeSkip bool
-
-// filterSmudge is a gateway to the `smudge()` function and serves to bail out
-// immediately if the pointer decoded from "from" has no data (i.e., is empty).
-// This function, unlike the implementation found in the legacy smudge command,
-// only combines the `io.Reader`s when necessary, since the implementation
-// found in `*git.PacketReader` blocks while waiting for the following packet.
-func filterSmudge(to io.Writer, from io.Reader, filename string, skip bool, filter *filepathfilter.Filter) error {
-	var pbuf bytes.Buffer
-	from = io.TeeReader(from, &pbuf)
-
-	ptr, err := lfs.DecodePointer(from)
-	if err != nil {
-		// If we tried to decode a pointer out of the data given to us,
-		// and the file was _empty_, write out an empty file in
-		// response. This occurs because when the clean filter
-		// encounters an empty file, and writes out an empty file,
-		// instead of a pointer.
-		if pbuf.Len() == 0 {
-			if _, cerr := io.Copy(to, &pbuf); cerr != nil {
-				Panic(cerr, "Error writing data to stdout:")
-			}
-			return nil
-		}
-
-		return err
-	}
-
-	return smudge(to, ptr, filename, skip, filter)
-}
 
 func filterCommand(cmd *cobra.Command, args []string) {
 	requireStdin("This command should be run by the Git filter process")
@@ -89,7 +59,7 @@ func filterCommand(cmd *cobra.Command, args []string) {
 			err = clean(w, req.Payload, req.Header["pathname"])
 		case "smudge":
 			w = git.NewPktlineWriter(os.Stdout, smudgeFilterBufferCapacity)
-			err = filterSmudge(w, req.Payload, req.Header["pathname"], skip, filter)
+			err = smudge(w, req.Payload, req.Header["pathname"], skip, filter)
 		default:
 			ExitWithError(fmt.Errorf("Unknown command %q", req.Header["command"]))
 		}

--- a/commands/command_filter_process.go
+++ b/commands/command_filter_process.go
@@ -4,7 +4,6 @@ import (
 	"fmt"
 	"io"
 	"os"
-	"strings"
 
 	"github.com/git-lfs/git-lfs/errors"
 	"github.com/git-lfs/git-lfs/filepathfilter"
@@ -85,7 +84,9 @@ func filterCommand(cmd *cobra.Command, args []string) {
 
 	if len(malformed) > 0 {
 		fmt.Fprintf(os.Stderr, "Encountered %d files that should have been pointers, but weren't:\n", len(malformed))
-		fmt.Fprintf(os.Stderr, "%s\n", strings.Join(malformed, ", "))
+		for _, m := range malformed {
+			fmt.Fprintf(os.Stderr, "\t%s\n", m)
+		}
 	}
 
 	if err := s.Err(); err != nil && err != io.EOF {

--- a/commands/command_smudge.go
+++ b/commands/command_smudge.go
@@ -1,7 +1,6 @@
 package commands
 
 import (
-	"bytes"
 	"fmt"
 	"io"
 	"os"
@@ -29,14 +28,9 @@ var (
 // were non-fatal, otherwise execution will halt and the process will be
 // terminated by using the `commands.Panic()` func.
 func smudge(to io.Writer, from io.Reader, filename string, skip bool, filter *filepathfilter.Filter) error {
-	pbuf, more, ptr, perr := lfs.DecodeFromHasMore(from)
+	ptr, pbuf, perr := lfs.DecodeFrom(from)
 	if perr != nil {
-		var r io.Reader = bytes.NewReader(pbuf)
-		if more {
-			r = io.MultiReader(r, from)
-		}
-
-		if _, err := io.Copy(to, r); err != nil {
+		if _, err := io.Copy(to, pbuf); err != nil {
 			return errors.Wrap(err, perr.Error())
 		}
 

--- a/commands/command_smudge.go
+++ b/commands/command_smudge.go
@@ -2,6 +2,7 @@ package commands
 
 import (
 	"bytes"
+	"fmt"
 	"io"
 	"os"
 
@@ -39,7 +40,9 @@ func smudge(to io.Writer, from io.Reader, filename string, skip bool, filter *fi
 			return errors.Wrap(err, perr.Error())
 		}
 
-		return nil
+		return errors.NewNotAPointerError(errors.Errorf(
+			"Unable to parse pointer at: %q", filename,
+		))
 	}
 
 	lfs.LinkOrCopyFromReference(ptr.Oid, ptr.Size)
@@ -82,7 +85,11 @@ func smudgeCommand(cmd *cobra.Command, args []string) {
 	filter := filepathfilter.New(cfg.FetchIncludePaths(), cfg.FetchExcludePaths())
 
 	if err := smudge(os.Stdout, os.Stdin, smudgeFilename(args), smudgeSkip, filter); err != nil {
-		Error(err.Error())
+		if errors.IsNotAPointerError(err) {
+			fmt.Fprintln(os.Stderr, err.Error())
+		} else {
+			Error(err.Error())
+		}
 	}
 }
 

--- a/commands/command_smudge.go
+++ b/commands/command_smudge.go
@@ -4,7 +4,6 @@ import (
 	"bytes"
 	"io"
 	"os"
-	"path/filepath"
 
 	"github.com/git-lfs/git-lfs/errors"
 	"github.com/git-lfs/git-lfs/filepathfilter"
@@ -28,7 +27,21 @@ var (
 // Any errors encountered along the way will be returned immediately if they
 // were non-fatal, otherwise execution will halt and the process will be
 // terminated by using the `commands.Panic()` func.
-func smudge(to io.Writer, ptr *lfs.Pointer, filename string, skip bool, filter *filepathfilter.Filter) error {
+func smudge(to io.Writer, from io.Reader, filename string, skip bool, filter *filepathfilter.Filter) error {
+	pbuf, more, ptr, perr := lfs.DecodeFromHasMore(from)
+	if perr != nil {
+		var r io.Reader = bytes.NewReader(pbuf)
+		if more {
+			r = io.MultiReader(r, from)
+		}
+
+		if _, err := io.Copy(to, r); err != nil {
+			return errors.Wrap(err, perr.Error())
+		}
+
+		return nil
+	}
+
 	lfs.LinkOrCopyFromReference(ptr.Oid, ptr.Size)
 	cb, file, err := lfs.CopyCallbackFile("smudge", filename, 1, 1)
 	if err != nil {
@@ -63,37 +76,20 @@ func smudgeCommand(cmd *cobra.Command, args []string) {
 	requireStdin("This command should be run by the Git 'smudge' filter")
 	lfs.InstallHooks(false)
 
-	// keeps the initial buffer from lfs.DecodePointer
-	b := &bytes.Buffer{}
-	r := io.TeeReader(os.Stdin, b)
-
-	ptr, perr := lfs.DecodePointer(r)
-	if perr != nil {
-		mr := io.MultiReader(b, os.Stdin)
-		if _, err := io.Copy(os.Stdout, mr); err != nil {
-			Panic(err, "Error writing data to stdout:")
-		}
-		return
-	}
-
 	if !smudgeSkip && cfg.Os.Bool("GIT_LFS_SKIP_SMUDGE", false) {
 		smudgeSkip = true
 	}
 	filter := filepathfilter.New(cfg.FetchIncludePaths(), cfg.FetchExcludePaths())
-	if err := smudge(os.Stdout, ptr, smudgeFilename(args, perr), smudgeSkip, filter); err != nil {
+
+	if err := smudge(os.Stdout, os.Stdin, smudgeFilename(args), smudgeSkip, filter); err != nil {
 		Error(err.Error())
 	}
 }
 
-func smudgeFilename(args []string, err error) string {
+func smudgeFilename(args []string) string {
 	if len(args) > 0 {
 		return args[0]
 	}
-
-	if errors.IsSmudgeError(err) {
-		return filepath.Base(errors.GetContext(err, "FileName").(string))
-	}
-
 	return "<unknown file>"
 }
 

--- a/lfs/pointer.go
+++ b/lfs/pointer.go
@@ -104,33 +104,32 @@ func DecodePointerFromFile(file string) (*Pointer, error) {
 	return DecodePointer(f)
 }
 func DecodePointer(reader io.Reader) (*Pointer, error) {
-	_, p, err := DecodeFrom(reader)
+	p, _, err := DecodeFrom(reader)
 	return p, err
 }
 
-func DecodeFrom(reader io.Reader) ([]byte, *Pointer, error) {
-	output, _, p, err := DecodeFromHasMore(reader)
-	return output, p, err
-}
-
-// DecodeFromHasMore decodes an *lfs.Pointer from the given io.Reader, "reader".
+// DecodeFrom decodes an *lfs.Pointer from the given io.Reader, "reader".
 // If the pointer encoded in the reader could successfully be read and decoded,
 // it will be returned with a nil error.
 //
-// If the pointer could not be decoded, a buffer of the data read so far, along
-// with whether or not there was more data not yet consumed will be returned
-// instead.
-func DecodeFromHasMore(reader io.Reader) ([]byte, bool, *Pointer, error) {
+// If the pointer could not be decoded, an io.Reader containing the entire
+// blob's data will be returned, along with a parse error.
+func DecodeFrom(reader io.Reader) (*Pointer, io.Reader, error) {
 	buf := make([]byte, blobSizeCutoff)
 	written, rerr := reader.Read(buf)
 	output := buf[0:written]
 
+	var buffer io.Reader = bytes.NewReader(output)
+	if rerr != io.EOF {
+		buffer = io.MultiReader(buffer, reader)
+	}
+
 	if rerr != nil && rerr != io.EOF {
-		return output, true, nil, rerr
+		return nil, buffer, rerr
 	}
 
 	p, err := decodeKV(bytes.TrimSpace(output))
-	return output, rerr != io.EOF, p, err
+	return p, buffer, err
 }
 
 func verifyVersion(version string) error {

--- a/lfs/pointer.go
+++ b/lfs/pointer.go
@@ -116,20 +116,20 @@ func DecodePointer(reader io.Reader) (*Pointer, error) {
 // blob's data will be returned, along with a parse error.
 func DecodeFrom(reader io.Reader) (*Pointer, io.Reader, error) {
 	buf := make([]byte, blobSizeCutoff)
-	written, rerr := reader.Read(buf)
-	output := buf[0:written]
+	n, err := reader.Read(buf)
+	buf = buf[:n]
 
-	var buffer io.Reader = bytes.NewReader(output)
-	if rerr != io.EOF {
-		buffer = io.MultiReader(buffer, reader)
+	var contents io.Reader = bytes.NewReader(buf)
+	if err != io.EOF {
+		contents = io.MultiReader(contents, reader)
 	}
 
-	if rerr != nil && rerr != io.EOF {
-		return nil, buffer, rerr
+	if err != nil && err != io.EOF {
+		return nil, contents, err
 	}
 
-	p, err := decodeKV(bytes.TrimSpace(output))
-	return p, buffer, err
+	p, err := decodeKV(bytes.TrimSpace(buf))
+	return p, contents, err
 }
 
 func verifyVersion(version string) error {

--- a/lfs/pointer.go
+++ b/lfs/pointer.go
@@ -109,16 +109,28 @@ func DecodePointer(reader io.Reader) (*Pointer, error) {
 }
 
 func DecodeFrom(reader io.Reader) ([]byte, *Pointer, error) {
+	output, _, p, err := DecodeFromHasMore(reader)
+	return output, p, err
+}
+
+// DecodeFromHasMore decodes an *lfs.Pointer from the given io.Reader, "reader".
+// If the pointer encoded in the reader could successfully be read and decoded,
+// it will be returned with a nil error.
+//
+// If the pointer could not be decoded, a buffer of the data read so far, along
+// with whether or not there was more data not yet consumed will be returned
+// instead.
+func DecodeFromHasMore(reader io.Reader) ([]byte, bool, *Pointer, error) {
 	buf := make([]byte, blobSizeCutoff)
-	written, err := reader.Read(buf)
+	written, rerr := reader.Read(buf)
 	output := buf[0:written]
 
-	if err != nil && err != io.EOF {
-		return output, nil, err
+	if rerr != nil && rerr != io.EOF {
+		return output, true, nil, rerr
 	}
 
 	p, err := decodeKV(bytes.TrimSpace(output))
-	return output, p, err
+	return output, rerr != io.EOF, p, err
 }
 
 func verifyVersion(version string) error {

--- a/lfs/pointer_clean.go
+++ b/lfs/pointer_clean.go
@@ -5,6 +5,7 @@ import (
 	"crypto/sha256"
 	"encoding/hex"
 	"io"
+	"io/ioutil"
 	"os"
 
 	"github.com/git-lfs/git-lfs/config"
@@ -76,8 +77,9 @@ func copyToTemp(reader io.Reader, fileSize int64, cb progress.CopyCallback) (oid
 		cb = nil
 	}
 
-	by, ptr, err := DecodeFrom(reader)
-	if err == nil && len(by) < 512 {
+	ptr, buf, err := DecodeFrom(reader)
+	by, rerr := ioutil.ReadAll(buf)
+	if rerr != nil || (err == nil && len(by) < 512) {
 		err = errors.NewCleanPointerError(ptr, by)
 		return
 	}

--- a/lfs/pointer_test.go
+++ b/lfs/pointer_test.go
@@ -3,6 +3,7 @@ package lfs
 import (
 	"bufio"
 	"bytes"
+	"io/ioutil"
 	"reflect"
 	"strings"
 	"testing"
@@ -168,11 +169,13 @@ size 12345`
 }
 
 func TestDecodeFromEmptyReader(t *testing.T) {
-	by, p, err := DecodeFrom(strings.NewReader(""))
+	p, buf, err := DecodeFrom(strings.NewReader(""))
+	by, rerr := ioutil.ReadAll(buf)
 
+	assert.Nil(t, rerr)
 	assert.EqualError(t, err, "Pointer file error: invalid header")
 	assert.Nil(t, p)
-	assert.Empty(t, string(by))
+	assert.Empty(t, by)
 }
 
 func TestDecodeInvalid(t *testing.T) {

--- a/test/test-filter-branch.sh
+++ b/test/test-filter-branch.sh
@@ -1,0 +1,41 @@
+#!/usr/bin/env bash
+
+. "test/testlib.sh"
+
+begin_test "filter-branch (git-lfs/git-lfs#1773)"
+(
+  set -e
+
+  reponame="filter-branch"
+  setup_remote_repo "$reponame"
+  clone_repo "$reponame" "$reponame"
+
+  contents_a="contents (a)"
+  printf "$contents_a" > a.dat
+  git add a.dat
+  git commit -m "add a.dat"
+
+  contents_b="contents (b)"
+  printf "$contents_b" > b.dat
+  git add b.dat
+  git commit -m "add b.dat"
+
+  contents_c="contents (c)"
+  printf "$contents_c" > c.dat
+  git add c.dat
+  git commit -m "add c.dat"
+
+  git filter-branch -f --prune-empty \
+    --tree-filter '
+      echo >&2 "---"
+      git rm --cached -r -q .
+      git lfs track "*.dat"
+      git add .
+    ' --tag-name-filter cat -- --all
+
+
+  assert_pointer "master" "a.dat" "$(calc_oid "$contents_a")" 12
+  assert_pointer "master" "b.dat" "$(calc_oid "$contents_b")" 12
+  assert_pointer "master" "c.dat" "$(calc_oid "$contents_c")" 12
+)
+end_test

--- a/test/test-malformed-pointers.sh
+++ b/test/test-malformed-pointers.sh
@@ -30,6 +30,10 @@ begin_test "malformed pointers"
   pushd .. >/dev/null
     clone_repo "$reponame" "$reponame-assert"
 
+    grep "malformed_small.dat" clone.log
+    grep "malformed_exact.dat" clone.log
+    grep "malformed_large.dat" clone.log
+
     expected_small="$(cat ../$reponame/malformed_small.dat)"
     expected_exact="$(cat ../$reponame/malformed_exact.dat)"
     expected_large="$(cat ../$reponame/malformed_large.dat)"

--- a/test/test-malformed-pointers.sh
+++ b/test/test-malformed-pointers.sh
@@ -20,7 +20,7 @@ begin_test "malformed pointers"
 
   git \
     -c "filter.lfs.process=" \
-    -c "filter.lfs.clean=" \
+    -c "filter.lfs.clean=cat" \
     -c "filter.lfs.required=false" \
     add *.dat
   git commit -m "add malformed pointer"

--- a/test/test-malformed-pointers.sh
+++ b/test/test-malformed-pointers.sh
@@ -1,0 +1,46 @@
+#!/usr/bin/env bash
+
+. "test/testlib.sh"
+
+begin_test "malformed pointers"
+(
+  set -e
+
+  reponame="malformed-pointers"
+  setup_remote_repo "$reponame"
+  clone_repo "$reponame" "$reponame"
+
+  git lfs track "*.dat"
+  git add .gitattributes
+  git commit -m "initial commit"
+
+  base64 /dev/urandom | head -c 1023 > malformed_small.dat
+  base64 /dev/urandom | head -c 1024 > malformed_exact.dat
+  base64 /dev/urandom | head -c 1025 > malformed_large.dat
+
+  git \
+    -c "filter.lfs.process=" \
+    -c "filter.lfs.clean=" \
+    -c "filter.lfs.required=false" \
+    add *.dat
+  git commit -m "add malformed pointer"
+
+  git push origin master
+
+  pushd .. >/dev/null
+    clone_repo "$reponame" "$reponame-assert"
+
+    expected_small="$(cat ../$reponame/malformed_small.dat)"
+    expected_exact="$(cat ../$reponame/malformed_exact.dat)"
+    expected_large="$(cat ../$reponame/malformed_large.dat)"
+
+    actual_small="$(cat malformed_small.dat)"
+    actual_exact="$(cat malformed_exact.dat)"
+    actual_large="$(cat malformed_large.dat)"
+
+    [ "$expected_small" = "$actual_small" ]
+    [ "$expected_exact" = "$actual_exact" ]
+    [ "$expected_large" = "$actual_large" ]
+  popd >/dev/null
+)
+end_test

